File: ElixirLeadge/lib/all_of_concepts/GenServerEx/user_manager.ex
defmodule User do
  defstruct [:id, :name, :email]
end

defmodule UserManager do
  use GenServer
  alias Jason, as: JSON
  @module __MODULE__
  @doc """
      API
  """
  def start_link(file_path) do
    GenServer.start(@module, file_path, name: @module)
  end

  def get_user(id) do
    GenServer.call(@module, {:get_user, id})
  end

  def add_user(user_model) do
    GenServer.cast(@module, {:add_user, user_model})
  end

  def update_user(user_id, update_model) do
    GenServer.cast(@module, {:update_user, user_id, update_model})
  end

  @doc """
      Callbacks for requests
  """
  @impl true
  def init(file_path) do
    state = load_users_from_file(file_path)
    {:ok, state}
  end

  @impl true
  def handle_cast({:add_user, user_model}, state) do
    updated_state = Map.put(state, user_model.id, user_model)
    {:noreply, updated_state}
  end

  @impl true
  def handle_cast({:update_user, id, update_model}, state) do
    case Map.get(state, id) do
      nil ->
        IO.puts("User with id #{id} not found")
        {:noreply, state}

      user ->
        updated_user = struct(user, update_model)
        updated_state = Map.put(state, id, updated_user)
        IO.puts("User with id #{id} updated successfully")
        {:noreply, updated_state}
    end
  end

  def read_file(file_path) do
#      IO.inspect(Path.expand(file_path))
    if File.exists?(file_path) do
      {:ok, content} = File.read(file_path)
      IO.inspect(content, label: "Json file content: ")
      {:ok, content}
    else
      {:error, "Json File doesn't exist"}
    end
  end

  def decode_json(file_content) do
    case JSON.decode(file_content) do
      {:ok, json_data} -> {:ok, json_data}
      {:error, reason} -> {:error, "Failed to decode JSOn file #{reason}"}
    end
  end

  defp process_users(data) do
      data
      |> Enum.map(fn %{"id" => id, "name" => name, "email" => email} ->
          %User{id: id, name: name, email: email}
      end)
      |> Enum.into(%{}, &{&1.id, &1})
  end

  # Function which load json content as User Models
  defp load_users_from_file(file_path) do
    with {:ok, content} <- read_file(file_path),
         {:ok, data} <- decode_json(content) do
      process_users(data)
    else
      {:error, reason} -> IO.puts("Error: #{reason}")
    end
  end
end

# running server
{:ok, _pid} = UserManager.start_link("static/users.json")
IO.inspect(UserManager.get_user(1))

File: Revise/Net/Erlang/example/src/Concurrency/user_manager.erl
-module(user_manager).
-author("golo").
-behaviour(gen_server).
%% API
-export([
  get_records_data/1,
  start_link/0,
  stop/1,
  add_user/1,
  remove_user/1,
  update_user/2,
  get_user/1,
  get_all_users/0
]).
-export([init/1,
  handle_call/3,
  handle_cast/2,
  handle_info/2
]).

-record(user, {id, name, age, email}).

%%  c(user_manager). user_manager:function().

init([]) ->
  {ok, []}.

get_records_data(Data) ->
  [Name, Age, Email] = [maps:get(Key, Data) || Key <- maps:keys(Data)],
  [Name, Age, Email].

%% Userdata -> map # name age email
handle_call({add_user, UserData}, _From, Users) ->
  [Name, Age, Email] = get_records_data(UserData),
  NewUser = #user{id= length(Users) + 1, name = Name, age = Age, email = Email },
  ChangedUsers = [NewUser | Users],
  {reply, ok, ChangedUsers};

handle_call({remove_user, UserId}, _From, Users) ->
  UsersWithoutRemovalId = lists:filter(
    fun(User) -> User#user.id =/= UserId end, Users
  ),
  {reply, ok, UsersWithoutRemovalId};

handle_call({update_user, UserId, NewData}, _From, Users) ->
    UpdatedUsers = lists:map(
      fun(User) ->
        if User#user.id =:= UserId ->
          [Name, Age, Email] = get_records_data(NewData),
          User#user{name = Name, age = Age, email = Email };
        true ->
          User
        end
      end, Users
    ),
  {reply, ok, UpdatedUsers};


handle_call({get_user, UserId}, _From, Users) ->
  SearchedUser = lists:filter(
    fun(User) -> User#user.id =:= UserId end, Users
  ),
  {reply, SearchedUser, Users};

handle_call(get_all_users, _From, Users) ->
  {reply, Users, Users};

handle_call({stop, Pid }, _From, _Users) ->
  {reply, "Pid " ++ Pid ++ "removed", []}.

handle_cast(_Msg, State) ->
  {noreply, State}.

handle_info(_Info, State) ->
  {noreply, State}.

%% API
start_link() ->
  gen_server:start(?MODULE, [], []).

stop(Pid) ->
  gen_server:call(?MODULE, {stop, Pid}).

add_user(NewUserData) ->
  gen_server:call(?MODULE, {add_user, NewUserData}).

remove_user(UserId) ->
  gen_server:call(?MODULE, {remove_user, UserId}).

update_user(UserId, NewData) ->
  gen_server:call(?MODULE, {update_user, UserId, NewData}).

get_user(UserId) ->
  gen_server:call(?MODULE, {get_user, UserId}).

get_all_users() ->
  gen_server:call(?MODULE, get_all_users).

